calico/felix/config.py

```diff
@@ -46,166 +46,311 @@
+        values before the rest of the config structure can be used.
+
         :raises EtcdException
         """
-        self._KnownSections = set()
-        self._KnownObjects  = set()
+        self.parameters = {}
+
+        self.add_parameter("EtcdAddr", "Address and port for etcd",
+                           "localhost:4001", sources = [ENV, FILE])
+        self.add_parameter("FelixHostname", "Felix compute host hostname",
+                           socket.gethostname(), sources = [ENV, FILE])
+
+        self.add_parameter("StartupCleanupDelay", "Delay before cleanup starts",
+                           30, value_is_int=True)
+        self.add_parameter("MetadataAddr", "Metadata IP address or hostname",
+                           "127.0.0.1")
+        self.add_parameter("MetadataPort", "Metadata Port",
+                           8775, value_is_int=True)
+        self.add_parameter("InterfacePrefix", "Interface name prefix", None)
+        self.add_parameter("LogFilePath",
+                           "Path to log file", "/var/log/calico/felix.log")
+        self.add_parameter("LogSeverityFile",
+                           "Log severity for logging to file", "INFO")
+        self.add_parameter("LogSeveritySys",
+                           "Log severity for logging to syslog", "ERROR")
+        self.add_parameter("LogSeverityScreen",
+                           "Log severity for logging to screen", "ERROR")
+
+        # Read the environment variables, then the configuration file.
+        self._read_env_vars()
+        self._read_cfg_file(config_path)
+        self._finish_update(final=False)
+
+    def add_parameter(self, name, description, default, **kwargs):
+        """
+        Put a parameter in the parameter dictionary.
+        """
+        self.parameters[name] = ConfigParameter(
+            name, description, default, **kwargs)
 
-        self._config_path = config_path
-        self._items = {}
+    def _finish_update(self, final=False):
+        """
+        Config has been completely read. Called twice - once after reading from
+        environment and config file (so we should be able to access etcd), and
+        once after reading from etcd (so we have all the config ready to go).
+
+        Responsible for :
+        - storing the parameters in the relevant fields in the structure
+        - validating the configuration is valid (for this stage in the process)
+        - updating logging parameters (if this is the final call)
+        :param final: Have we completed (rather than just read env and config file)
+        """
+        self.ETCD_ADDR = self.parameters["EtcdAddr"].value
+        self.HOSTNAME = self.parameters["FelixHostname"].value
+        self.STARTUP_CLEANUP_DELAY = self.parameters["StartupCleanupDelay"].value
+        self.METADATA_IP = self.parameters["MetadataAddr"].value
+        self.METADATA_PORT = self.parameters["MetadataPort"].value
```

Comment:
It does - for integer values, the value is stored as an int (see the "set" method).